openingd/dualopend.c

```diff
@@ -95,8 +123,1303 @@ struct state {
+							    NULL, false);
+			}
+			negotiation_aborted(state, am_opener,
+					    tal_fmt(tmpctx, "They sent error %s",
+						    err));
+			/* Return NULL so caller knows to stop negotiating. */
+			return NULL;
+		}
+
+		/*~ We do not support multiple "live" channels, though the
+		 * protocol has a "channel_id" field in all non-gossip messages
+		 * so it's possible.  Our one-process-one-channel mechanism
+		 * keeps things simple: if we wanted to change this, we would
+		 * probably be best with another daemon to de-multiplex them;
+		 * this could be connectd itself, in fact. */
+		if (is_wrong_channel(msg, &state->channel_id, &actual)) {
+			status_debug("Rejecting %s for unknown channel_id %s",
+				     wire_type_name(fromwire_peektype(msg)),
+				     type_to_string(tmpctx, struct channel_id,
+						    &actual));
+			sync_crypto_write(state->pps,
+					  take(towire_errorfmt(NULL, &actual,
+							       "Multiple channels"
+							       " unsupported")));
+			tal_free(msg);
+			continue;
+		}
+
+		/* If we get here, it's an interesting message. */
+		return msg;
+	}
+}
+
+static void run_tx_interactive(struct state *state, struct wally_psbt **orig_psbt)
+{
+	/* Opener always sends the first utxo info */
+	bool we_complete = false, they_complete = false;
+	u8 *msg;
+	struct wally_psbt *psbt = *orig_psbt;
+
+	while (!(we_complete && they_complete)) {
+		struct channel_id cid;
+		enum wire_type t;
+		u16 serial_id;
+
+		/* Reset their_complete to false every round,
+		 * they have to re-affirm every time  */
+		they_complete = false;
+
+		msg = sync_crypto_read(tmpctx, state->pps);
+		t = fromwire_peektype(msg);
+		switch (t) {
+		case WIRE_TX_ADD_INPUT: {
+			const u8 *tx_bytes, *redeemscript;
+			u16 max_witness_len;
+			u32 outnum, sequence;
+			size_t len;
+			struct bitcoin_tx *tx;
+			struct bitcoin_txid txid;
+			struct amount_sat amt;
+			struct tlv_tx_add_input_tlvs *add_tlvs =
+				tlv_tx_add_input_tlvs_new(tmpctx);
+
+			if (!fromwire_tx_add_input(tmpctx, msg, &cid,
+						   &serial_id,
+						   cast_const2(u8 **, &tx_bytes),
+						   &outnum, &sequence,
+						   &max_witness_len,
+						   cast_const2(u8 **, &redeemscript),
+						   add_tlvs))
+				peer_failed(state->pps, &state->channel_id,
+					    "Parsing tx_add_input %s",
+					    tal_hex(tmpctx, msg));
+
+			check_channel_id(state, &cid, &state->channel_id);
+			/*
+			 * BOLT-FIXME #2:
+			 * - if is the `initiator`:
+			 *   - MUST send even `serial_id`s
+			 * - MUST fail the transaction collaboration if:
+			 *   ...
+			 * - it receives a `serial_id` from the peer
+			 *   with the incorrect parity
+			 */
+			if (serial_id % 2 != 0)
+				peer_failed(state->pps, &state->channel_id,
+					    "Invalid serial_id rcvd. %u", serial_id);
+			/*
+			 * BOLT-FIXME #2:
+			 * - MUST fail the transaction collaboration if:
+			 *   ...
+			 *  - it recieves a duplicate `serial_id`
+			 */
+			if (psbt_has_serial_input(psbt, serial_id))
```

Comment:
Instead, implement psbt_find_serial_input() which returns -1 if not found, otherwise an index.

This makes remove much simpler, too.

And similarly for _output.